### PR TITLE
Mention maxSurge and maxUnavailable parameters in DaemonSet concept

### PR DESCRIPTION
**Pull request by tanayvaswani**
fix: [#41440](https://github.com/kubernetes/website/issues/41440)

Page modified: [DaemonSet's concept page](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/)

Issue: `maxSurge` parameter was not documented under the DaemonSet concept. 

Solution: Added the brief info about the `maxSurge` and `maxUnavailable` parameters under the DaemonSet's concept page along with some of the yaml code defining the use case of this particular parameter. 

/language en

### CODE REVIEW

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 1]
```diff
@@ -201,6 +201,31 @@ them according to its `updateStrategy`.
...
+
+Suppose the value of `maxSurge` is set to 1 and the desired count is 5, Kubernetes can create up to 6 pods while performing the update process. This parameter might be useful when you want to temporarily upscale the number of pods while performing the update.
+
+### MaxUnavailable
+
+This particular parameter defines the maximum number of pods that can be unavailable while performing the update. In simpler terms, they are being deleted and not yet replaced by new pods, during the update process. `maxUnavailable` can specify the number or percentage of pods that can be available at that particular moment. Let's look at an example below: 
+
+Suppose the value of `maxUnavailable` is set to 1 and the desired count is 5, Kubernetes can delete and replace pods one at a time, ensuring the fact that there should always be at least 4 pods available. This parameter might be useful to maintain the desired level of availability during updates.
+
+You may want to set `.spec.updateStrategy.rollingUpdate.maxUnavailable` (default to 1), `.spec.minReadySeconds` (default to 0) and `.spec.updateStrategy.rollingUpdate.maxSurge` (defaults to 0) as well.
```
Comment (Ritikaa96) [inline content/en/docs/concepts/workloads/controllers/daemonset.md:218]:
```suggestion
You need to set `.spec.updateStrategy.rollingUpdate.maxUnavailable` (default to 1), `.spec.minReadySeconds` (default to 0) and `.spec.updateStrategy.rollingUpdate.maxSurge` (defaults to 0).
Please note that the default value of `maxSurge` and `maxUnavailable` is 25% of `replicas`.
```

**Review by Ritikaa96 — CHANGES_REQUESTED** [review]

**Comment by tanayvaswani** [comment]
Thanks @Ritikaa96 :)

**Comment by sftim** [comment]
/retitle Mention maxSurge and maxUnavailable parameters in DaemonSet concept
/sig apps

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 2]
```diff
@@ -201,6 +201,39 @@ them according to its `updateStrategy`.
...
+`maxSurge` helps you to regulate the surge of additional pods beyond the desired count. 
+Let's look at an example below: 
+Suppose the value of `maxSurge` is set to 1 and the desired count is 5, Kubernetes can create up to 6 pods while performing the update process.
+This parameter is useful when you want to temporarily upscale the number of pods while performing the update.
+
+### MaxUnavailable
+
+This particular parameter defines the maximum number of pods that can be unavailable while performing the update.
+In simpler terms, they are being deleted and not yet replaced by new pods, during the update process. 
+`maxUnavailable` can specify the number or percentage of pods that can be available at that particular moment. Let's look at an example below:
```
Comment (sftim) [inline content/en/docs/concepts/workloads/controllers/daemonset.md:219]:
```suggestion
`maxUnavailable` specifies either the maximum number or maximum proportion of pods that can
be (voluntarily) [disrupted](/docs/concepts/workloads/pods/disruptions/#voluntary-and-involuntary-disruptions)
at any particular moment during a rollout.

#### maxUnavailable and Pod failures

During a DaemonSet rollout, one of the old pods maybe become unavailable for some reason
(for example: the `Ready` condition transitions to false, the  Pod is evicted, or the node is drained).
If that occurs, the DaemonSet controller immediately creates a replacement Pod on that node,
and avoids removing other old Pods until the number of unavailable Pods fits inside the
`maxUnavailable` threshold.

### Example {#updating-a-daemonset-example}
```

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 3]
```diff
@@ -201,6 +201,39 @@ them according to its `updateStrategy`.
...
+This parameter is responsible for determining the maximum number of pods that can be created above the desired number of pods mentioned in the DaemonSet.
+While performing a rolling update, Kubernetes will keep creating new pods to replace the old ones. 
+`maxSurge` helps you to regulate the surge of additional pods beyond the desired count. 
+Let's look at an example below: 
+Suppose the value of `maxSurge` is set to 1 and the desired count is 5, Kubernetes can create up to 6 pods while performing the update process.
+This parameter is useful when you want to temporarily upscale the number of pods while performing the update.
+
+### MaxUnavailable
+
+This particular parameter defines the maximum number of pods that can be unavailable while performing the update.
```
Comment (sftim) [inline content/en/docs/concepts/workloads/controllers/daemonset.md:217]:
```suggestion
This parameter defines the maximum number of pods, belonging to a DaemonSet,
that can be unavailable while performing an update.
```

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 4]
```diff
@@ -201,6 +201,39 @@ them according to its `updateStrategy`.
 
 You can [perform a rolling update](/docs/tasks/manage-daemon/update-daemon-set/) on a DaemonSet.
 
+While working with the rolling updates of the DaemonSets, you will come across two important parameters: `maxSurge` and `maxUnavailable`.
+
+### MaxSurge 
+
+This parameter is responsible for determining the maximum number of pods that can be created above the desired number of pods mentioned in the DaemonSet.
```
Comment (sftim) [inline content/en/docs/concepts/workloads/controllers/daemonset.md:208]: Is this true?
```suggestion
If you set this parameter to a number above 0, you allow the DaemonSet controller
to create temporary duplicate Pods during a rollout.
Because a DaemonSet aims to have one healthy Pod per node, this field also implies
the number of _nodes_ that can have a duplicate Pod on them at any one moment
during the rollout.
```

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 5]
```diff
@@ -201,6 +201,39 @@ them according to its `updateStrategy`.
...
+While performing a rolling update, Kubernetes will keep creating new pods to replace the old ones. 
+`maxSurge` helps you to regulate the surge of additional pods beyond the desired count. 
+Let's look at an example below: 
+Suppose the value of `maxSurge` is set to 1 and the desired count is 5, Kubernetes can create up to 6 pods while performing the update process.
+This parameter is useful when you want to temporarily upscale the number of pods while performing the update.
```
Comment (sftim) [inline content/en/docs/concepts/workloads/controllers/daemonset.md:213]:
```suggestion

#### maxSurge and Pod failures

During a DaemonSet rollout, one of the old pods maybe become unavailable for some reason
(for example: the `Ready` condition transitions to false, the  Pod is evicted, or the node is drained).
If that occurs, the DaemonSet controller immediately creates a replacement Pod on that node,
without considering surge limits. The replacement Pod uses the current (new) Pod template, not
the previous one.
```

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 6]
```diff
@@ -201,6 +201,39 @@ them according to its `updateStrategy`.
...
+
+This particular parameter defines the maximum number of pods that can be unavailable while performing the update.
+In simpler terms, they are being deleted and not yet replaced by new pods, during the update process. 
+`maxUnavailable` can specify the number or percentage of pods that can be available at that particular moment. Let's look at an example below: 
+
+Suppose the value of `maxUnavailable` is set to 1 and the desired count is 5,
+Kubernetes can delete and replace pods one at a time, ensuring the fact that there should always be at least 4 pods available.
+This parameter is useful to maintain the desired level of availability during updates.
+
+You need to set `.spec.updateStrategy.rollingUpdate.maxUnavailable` (default to 1), `.spec.minReadySeconds` (default to 0) and `.spec.updateStrategy.rollingUpdate.maxSurge` (defaults to 0).
```
[the remark this answers is not in the document]
Comment (tanayvaswani) [reply content/en/docs/concepts/workloads/controllers/daemonset.md:225]: It will use the default values imo.

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 7]
```diff
@@ -201,6 +201,39 @@ them according to its `updateStrategy`.
...
+Please note that the default value of `maxSurge` and `maxUnavailable` is 25% of `replicas`.
```
[the remark this answers is not in the document]
Comment (Ritikaa96) [reply content/en/docs/concepts/workloads/controllers/daemonset.md]: Hi @sftim @tanayvaswani [here](https://github.com/kubernetes/kubernetes/blob/af99df6a6bead927d61746dc431fc1b526af55a0/pkg/apis/apps/types.go#L497) the code file says: **By default, a value of 1 is used.**
```go
// The maximum number of pods that can be scheduled above the original number of
	// pods.
	// Value can be an absolute number (ex: 5) or a percentage of total pods at
	// the start of the update (ex: 10%). This can not be 0 if MaxUnavailable is 0.
	// Absolute number is calculated from percentage by rounding up.
	// By default, a value of 1 is used.
	// Example: when this is set to 30%, the new RC can be scaled up by 30%
	// immediately when the rolling update starts. Once old pods have been killed,
	// new RC can be scaled up further, ensuring that total number of pods running
	// at any time during the update is at most 130% of original pods.
	// +optional
	MaxSurge intstr.IntOrString
``` 
Is this stated right?

[the remark this answers is not in the document]
Comment (Ritikaa96) [reply content/en/docs/concepts/workloads/controllers/daemonset.md]: I tested on my system for deployment, and the `kubectl get deployment nginx-deployment -oyaml` command shows maxSurge and maxUnavailable to be 25% when we don't state them explicitly.
for more reference : https://github.com/kubernetes/kubernetes/blob/af99df6a6bead927d61746dc431fc1b526af55a0/pkg/apis/apps/v1/defaults.go#L59

```go
	if strategy.RollingUpdate.MaxSurge == nil {
			// Set default MaxSurge as 25% by default.
			maxSurge := intstr.FromString("25%")
			strategy.RollingUpdate.MaxSurge = &maxSurge
		}
```

File: content/en/docs/concepts/workloads/controllers/daemonset.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (sftim) [reply content/en/docs/concepts/workloads/controllers/daemonset.md:247]: I don't think you need to set then, then.

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 8]
```diff
@@ -201,6 +201,63 @@ them according to its `updateStrategy`.
...
+
+#### maxSurge and Pod failures
+
+During a DaemonSet rollout, one of the old pods maybe become unavailable for some reason
+(for example: the `Ready` condition transitions to false, the  Pod is evicted, or the node is drained).
+If that occurs, the DaemonSet controller immediately creates a replacement Pod on that node,
+without considering surge limits. The replacement Pod uses the current (new) Pod template, not
+the previous one.
+
+### MaxUnavailable
```
Comment (sftim) [inline content/en/docs/concepts/workloads/controllers/daemonset.md:224]:
```suggestion
### maxUnavailable
```

File: content/en/docs/concepts/workloads/controllers/daemonset.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (Ritikaa96) [reply content/en/docs/concepts/workloads/controllers/daemonset.md]: In [API Overview](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#rollingupdatedaemonset-v1-apps) it is stated clearly. 
```
This can not be 0 if MaxUnavailable is 0. Absolute number is calculated from percentage by rounding up to a minimum of 1. Default value is 0. Example: when this is set to 30%, at most 30% of the total number of nodes that should be running the daemon pod (i.e. status.desiredNumberScheduled) can have their a new pod created before the old pod is marked as deleted.
```

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 9]
```diff
@@ -201,6 +201,63 @@ them according to its `updateStrategy`.
...
+If you set this parameter to a number above 0, you allow the DaemonSet controller
+to create temporary duplicate Pods during a rollout.
+Because a DaemonSet aims to have one healthy Pod per node, this field also implies
+the number of _nodes_ that can have a duplicate Pod on them at any one moment
+during the rollout.
+While performing a rolling update, Kubernetes will keep creating new pods to replace the old ones. 
+`maxSurge` helps you to regulate the surge of additional pods beyond the desired count. 
+
+#### maxSurge and Pod failures
+
```
Comment (atiratree) [inline content/en/docs/concepts/workloads/controllers/daemonset.md:217]: I don't know if is useful to specify this for both `maxSurge` and `maxUnavailable`. This behaviour will occurr in all circumstances as both of the constraints cannot be 0. I think it is best to remove it or move this to general section just called `Pod failures`.

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 10]
```diff
@@ -201,6 +201,63 @@ them according to its `updateStrategy`.
 
 You can [perform a rolling update](/docs/tasks/manage-daemon/update-daemon-set/) on a DaemonSet.
 
+While working with the rolling updates of the DaemonSets, you will come across two important parameters: `maxSurge` and `maxUnavailable`.
+
+### maxSurge
+
+If you set this parameter to a number above 0, you allow the DaemonSet controller
+to create temporary duplicate Pods during a rollout.
+Because a DaemonSet aims to have one healthy Pod per node, this field also implies
```
Comment (atiratree) [inline content/en/docs/concepts/workloads/controllers/daemonset.md:210]:
```suggestion
Because a DaemonSet aims to have one healthy Pod per schedulable node, this field also implies
```

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 11]
```diff
@@ -201,6 +201,63 @@ them according to its `updateStrategy`.
...
+the number of _nodes_ that can have a duplicate Pod on them at any one moment
+during the rollout.
+While performing a rolling update, Kubernetes will keep creating new pods to replace the old ones. 
+`maxSurge` helps you to regulate the surge of additional pods beyond the desired count.
```
Comment (atiratree) [inline content/en/docs/concepts/workloads/controllers/daemonset.md:214]: please mention that maxSurge cannot be always used, if the pod requires a specific resource (eg hostPort) https://kubernetes.io/blog/2022/09/15/app-rollout-features-reach-stable/#how-use-daemonset-maxsurge. The application must allow 2 pods per node.

Since maxUnavailable is probably safer. I would mention it first and then follwoup with describing maxSurge.

Comment (atiratree) [reply content/en/docs/concepts/workloads/controllers/daemonset.md:214]: also this part from the blog is useful: 

> Kubernetes system-level components like CNI, CSI are typically run as DaemonSets. These components can have impact on the availability of the workloads if those DaemonSets go down momentarily during the upgrades. The feature allows DaemonSet pods to temporarily increase their number, thereby ensuring zero-downtime for the DaemonSets.

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 12]
```diff
@@ -201,6 +201,63 @@ them according to its `updateStrategy`.
...
+(for example: the `Ready` condition transitions to false, the  Pod is evicted, or the node is drained).
+If that occurs, the DaemonSet controller immediately creates a replacement Pod on that node,
+without considering surge limits. The replacement Pod uses the current (new) Pod template, not
+the previous one.
+
+### maxUnavailable
+
+This parameter defines the maximum number of pods, belonging to a DaemonSet,
+that can be unavailable while performing an update.
+In simpler terms, they are being deleted and not yet replaced by new pods, during the update process.
```
Comment (atiratree) [inline content/en/docs/concepts/workloads/controllers/daemonset.md:228]: hm, I think this can be a bit confusing. There can be a slight delay between the deletion and replacement, but practically it should be unnoticeable. I think it is safe to say that the pods are replaced immediately, but they might take time to become available. that is app becomes `ready` + `minReadySeconds`

File: content/en/docs/concepts/workloads/controllers/daemonset.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (atiratree) [reply content/en/docs/concepts/workloads/controllers/daemonset.md:247]: maybe we can mention the default and constraints by each feature (take a look at the type definition) and then we just say what we are setting in the example?

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 13]
```diff
@@ -201,6 +201,63 @@ them according to its `updateStrategy`.
...
+their pods stopped for an update at any given time.
+
+```yaml
+spec:
+  updateStrategy:
+      type: RollingUpdate
+      rollingUpdate:
+        maxSurge: 1
+        maxUnavailable: 1
+```
```
Comment (atiratree) [inline content/en/docs/concepts/workloads/controllers/daemonset.md:259]: and then maybe we can describe a bit how the deployment will behave with regards to these features?

File: content/en/docs/concepts/workloads/controllers/daemonset.md (hunk 10 above)
Comment (sftim) [reply content/en/docs/concepts/workloads/controllers/daemonset.md:210]: Or “selected”? Were you thinking of not scheduling to nodes with taints that the DaemonSet doesn't tolerate, or about the node selector for the DaemonSet's spec @atiratree ?

File: content/en/docs/concepts/workloads/controllers/daemonset.md (hunk 11 above)
Comment (sftim) [reply content/en/docs/concepts/workloads/controllers/daemonset.md:214]: If a Pod _is_ compatible with `maxSurge`, which approach would we typically recommend?

**Comment by sftim** [comment]
@tanayvaswani once you've incorporated feedback, please:
- push your changes
- rebase against _main_
- [force] push your changes once more

File: content/en/docs/concepts/workloads/controllers/daemonset.md (hunk 11 above)
Comment (atiratree) [reply content/en/docs/concepts/workloads/controllers/daemonset.md:214]: this is the same choice people have to make when working with deployments. It is a tradeoff between resource cost vs availability and speed. If you want a quick rollout you set appropriate `maxSurge`. And if you want to bring down the resource usage and don't mind loosing availability on each node during a rollout, you set `maxUnavalable`.

Comment (sftim) [reply content/en/docs/concepts/workloads/controllers/daemonset.md:214]: It's a similar choice, but there's a much bigger chance of a conflict because the daemonset tries to provide a node-level service, and is implemented with the expectation that there's no duplicate process / container / pod. We need to account for that difference when we explain things.

Comment (atiratree) [reply content/en/docs/concepts/workloads/controllers/daemonset.md:214]: yes, in general `maxSurge` is not recommendable, because the node-level service has to be aware of it

File: content/en/docs/concepts/workloads/controllers/daemonset.md (hunk 12 above)
Comment (atiratree) [reply content/en/docs/concepts/workloads/controllers/daemonset.md:228]: It depends on how you define replacement:
1. is it just the Pod API object, even if it is no picked up by scheduler yet?
2. is it Pod after being assigned and running?
3. is it available Pod?

If the user expects 1., then it could be confusing. I think it would be better to describe in more detail to prevent the confusion.

```suggestion
During the upgrade process, the pods are being deleted, and it may take some time for the replacement pods to become available on each node.
```

or something similar in this spirit?

**Comment by sftim** [comment]
As this PR has not changed in a long time, I will close it.

@tanayvaswani you are welcome to reopen the PR and to update your branch with the appropriate fixes. If it feels more appropriate, you could start a new pull request.

/close
